CHANGELOG.md

```diff
@@ -1,12 +1,26 @@
 # Changelog
 
-All notable changes to this project will be documented in this file. The format
-is based on [Keep a Changelog].
+All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog].
 
 This project adheres to [Semantic Versioning].
 
 ## [Unreleased]
 
+## [4.0.0] - 2024-05-13
+
+Major rework of supported verbs and functionality.
+
+### Changed
+
+- Modify `upload` verb to enforce the following restrictions:
+  - When running in a CI environment, the value specified for the `--upload_token` option (or the `WALDO_UPLOAD_TOKEN` environment variable) _must_ be a CI token. Moreover, the `--app_id` option is _not allowed_.
+  - When running interactively (meaning _not_ in a CI environment), the value specified for the `--upload_token` option _must_ be an API token. If the `--upload_token` option is _not_ specified, the API token specified by the preceding `waldo auth` invocation (if any) is used. Moreover, the `--app_id` option is _required_.
```

Comment:
Thats very user-hostile behaviour and will need very VERY **VERY** good documentation around it because it matches the behaviour of no other tool on the market.

We shouldn't provide massively different behaviour depending on an environment variable, even if most CI tools set it.

Reply:
@vbfox @laurentsigal 

Very well then. I will remove the restrictions. The question then is: what is the proper order of precedence in determining the value for the upload token?

1. `--upload_token`
2. `WALDO_UPLOAD_TOKEN`
3. API token stashed by `waldo auth`

_OR_

1. `--upload_token`
2. API token stashed by `waldo auth`
3. `WALDO_UPLOAD_TOKEN`

???

Reply:
I'd say the first one, but debatable

1. Explicitly on the command
2. Implicitly in the (local) env or potentially passed explicitly at the same time as the command is called
3. Implicitly by a previous command saved in a global place

Reply:
I tend to agree. Also first one is less likely to break an existing script.